### PR TITLE
EZP-30983: Prevented users from reusing old password

### PR DESCRIPTION
**Pull request by adamwojs**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30983](https://jira.ez.no/browse/EZP-30983)
| **Bug/Improvement**| yes/no
| **New feature**    | yes/no
| **Target version** | `7.5`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Prevent users from reusing old (current) password when password expiration is enabled.

**TODO**:
- [X] Implement feature / fix a bug.
- [X] Implement tests.
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Repository/UserService.php [hunk 1]
```diff
@@ -1298,11 +1301,17 @@ public function validatePassword(string $password, PasswordValidationContext $co
...
+        if (isset($configuration['PasswordValueValidator'])) {
+            $errors = (new UserPasswordValidator($configuration['PasswordValueValidator']))->validate($password);
         }
 
-        return (new UserPasswordValidator($configuration['PasswordValueValidator']))->validate($password);
+        if ($context->user !== null &&
+            $this->getPasswordInfo($context->user)->hasExpirationDate() &&
+            $this->verifyPasswordForAPIUser($context->user->login, $password, $context->user)) {
+            $errors[] = new ValidationError('New password cannot be the same as old password', null, [], 'password');
+        }
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]:
```
        if ($context->user !== null &&
            $this->getPasswordInfo($context->user)->hasExpirationDate() &&
            $this->verifyPasswordForAPIUser($context->user->login, $password, $context->user) 
        ) {
            $errors[] = new ValidationError('New password cannot be the same as old password', null, [], 'password');
        }
```

File: eZ/Publish/Core/Repository/UserService.php [hunk 2]
```diff
@@ -1409,9 +1418,41 @@ private function getUserFieldDefinition(ContentType $contentType): ?FieldDefinit
...
     }
 
+    /**
+     * Verifies if the provided login and password are valid for eZ\Publish\SPI\Persistence\User.
+     *
+     * @param string $login User login
+     * @param string $password User password
+     * @param \eZ\Publish\SPI\Persistence\User $spiUser Loaded user handler
+     *
+     * @return bool return true if the login and password are sucessfully
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]:
```suggestion
     * @return bool return true if the login and password are sucessfully validate and false, if not.
```

File: eZ/Publish/Core/Repository/UserService.php [hunk 3]
```diff
@@ -1409,9 +1418,41 @@ private function getUserFieldDefinition(ContentType $contentType): ?FieldDefinit
...
+    }
+
+    /**
+     * Verifies if the provided login and password are valid for eZ\Publish\API\Repository\Values\User\User.
+     *
+     * @param string $login User login
+     * @param string $password User password
+     * @param \eZ\Publish\API\Repository\Values\User\User $apiUser Loaded user
+     *
+     * @return bool return true if the login and password are sucessfully
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]:
```suggestion
     * @return bool return true if the login and password are sucessfully validate and false, if not.
```

File: eZ/Publish/Core/Repository/UserService.php [hunk 4]
```diff
@@ -1420,24 +1461,38 @@ private function getUserFieldDefinition(ContentType $contentType): ?FieldDefinit
...
+
+    /**
+     * Verifies if the provided login and password are valid against given password hash and hash type.
+     *
+     * @param string $login User login
+     * @param string $plainPassword User password
+     * @param string $passwordHash User password hash
+     * @param int $hashAlgorithm Hash type
+     *
+     * @return bool return true if the login and password are sucessfully
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]: I think we have space for one-liner

File: eZ/Publish/Core/Repository/UserService.php [hunk 5]
```diff
@@ -1420,24 +1461,38 @@ private function getUserFieldDefinition(ContentType $contentType): ?FieldDefinit
...
+     * validate and false, if not.
+     */
+    private function doVerifyPassword(string $login, string $plainPassword, string $passwordHash, int $hashAlgorithm): bool
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]: this could be in separate lines

File: eZ/Publish/Core/Repository/UserService.php [hunk 6]
```diff
@@ -1420,24 +1461,38 @@ private function getUserFieldDefinition(ContentType $contentType): ?FieldDefinit
...
     {
         // In case of bcrypt let php's password functionality do it's magic
-        if ($spiUser->hashAlgorithm === APIUser::PASSWORD_HASH_BCRYPT ||
-            $spiUser->hashAlgorithm === APIUser::PASSWORD_HASH_PHP_DEFAULT) {
-            return password_verify($password, $spiUser->passwordHash);
+        if ($hashAlgorithm === APIUser::PASSWORD_HASH_BCRYPT ||
+            $hashAlgorithm === APIUser::PASSWORD_HASH_PHP_DEFAULT) {
+            return password_verify($plainPassword, $passwordHash);
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/UserService.php]:
```
        if ($hashAlgorithm === APIUser::PASSWORD_HASH_BCRYPT ||
            $hashAlgorithm === APIUser::PASSWORD_HASH_PHP_DEFAULT
        ) {
            return password_verify($plainPassword, $passwordHash);
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 7]
```diff
@@ -2873,7 +2873,31 @@ public function testValidatePassword(string $password, array $expectedErrorr)
...
+    {
+        $password = 'P@blish123!';
+
+        $userService = $this->getRepository()->getUserService();
+        // Password expiration needs to be enabled
+        $contentType = $this->createUserContentTypeWithPasswordExpirationDate();
+
+        $user = $this->createTestUserWithPassword($password, $contentType);
+
+        /* BEGIN: Use Case */
```
Comment (mikadamczyk) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 8]
```diff
@@ -2873,7 +2873,31 @@ public function testValidatePassword(string $password, array $expectedErrorr)
...
+        $context = new PasswordValidationContext([
+            'contentType' => $contentType,
+            'user' => $user,
+        ]);
+
+        $actualErrors = $userService->validatePassword($password, $context);
+        /* END: Use Case */
```
Comment (mikadamczyk) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 9]
```diff
@@ -2873,7 +2873,31 @@ public function testValidatePassword(string $password, array $expectedErrorr)
...
+
+        $this->assertEquals([
+            new ValidationError('New password cannot be the same as old password', null, [], 'password'),
+        ], $actualErrors);
```
Comment (mikadamczyk) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
        $this->assertEquals(
            [new ValidationError('New password cannot be the same as old password', null, [], 'password')], 
            $userService->validatePassword($password, $context)
        );
```
nitpick: without a temporary variable or with two of them

**Comment by lserwatka** [comment]
You can merge it up.

**Pull request merged**